[WPE] Implement theme color retrieval API

fredldotme opened the pull request:
#### 2246a2aa28d527e129ebec66183569db035fdc8a
<pre>
[WPE] Implement theme color retrieval API

<a href="https://bugs.webkit.org/show_bug.cgi?id=273747">https://bugs.webkit.org/show_bug.cgi?id=273747</a>

Reviewed by Adrian Perez de Castro and Carlos Garcia Campos.

Implements a property &quot;theme-color&quot; and a getter function webkit_web_view_get_theme_color()
for retrieving the web view&apos;s page theme color, with property changes for notification.

Combined changes:

* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
(WebKitWebViewClient::themeColorDidChange):
(webkitWebViewGetProperty):
(webkit_web_view_class_init):
* Source/WebKit/UIProcess/API/glib/WebKitWebView.h.in:
* Source/WebKit/UIProcess/API/wpe/APIViewClient.h:
(API::ViewClient::themeColorDidChange):
* Source/WebKit/UIProcess/API/wpe/PageClientImpl.cpp:
(WebKit::PageClientImpl::themeColorDidChange):
* Source/WebKit/UIProcess/API/wpe/PageClientImpl.h:
* Source/WebKit/UIProcess/API/wpe/WPEWebView.cpp:
(WKWPE::View::themeColorDidChange):
* Source/WebKit/UIProcess/API/wpe/WPEWebView.h:
* Source/WebKit/UIProcess/API/wpe/WebKitWebViewClient.h:
* Source/WebKit/UIProcess/API/wpe/WebKitWebViewWPE.cpp:
(webkit_web_view_get_theme_color):
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/LayerTreeHost.cpp:
(WebKit::LayerTreeHost::flushLayers):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebView.cpp:
(ThemeColorWebViewTest::themeColorChanged):
(ThemeColorWebViewTest::testThemeColorResult):
(ThemeColorWebViewTest::waitUntilThemeColorChanged):
(testWebViewThemeColor):
(beforeAll):

Canonical link: <a href="https://commits.webkit.org/296035@main">https://commits.webkit.org/296035@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/da8019d4a6faee74ff52ff9bfbff1fb3d6a48492

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/107125 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/26822 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/17221 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/112335 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/57672 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/109105 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/27493 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/35325 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/81325 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/110070 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/21820 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/96593 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/61687 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/21243 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/14721 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/57119 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/91171 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/14742 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/115428 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/34205 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/25206 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/90370 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/34579 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/92845 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/90103 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/35033 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/12809 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/29934 "The change is no longer eligible for processing.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17323 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/34130 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/39594 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/33877 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/37229 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/35523 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -238,6 +238,8 @@ enum {
     PROP_WEB_EXTENSION_MODE,
     PROP_DEFAULT_CONTENT_SECURITY_POLICY,
 
+    PROP_THEME_COLOR,
```
You may remove the empty line right above this one.

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -1168,6 +1174,11 @@ static void webkitWebViewGetProperty(GObject* object, guint propId, GValue* valu
         g_value_set_string(value, webkit_web_view_get_default_content_security_policy(webView));
         break;
+    case PROP_THEME_COLOR:
+        WebKitColor* color = new WebKitColor;
```
We need to be consistent in the memory allocator being used.

Looking at `Source/WebKit/UIProcess/API/wpe/WebKitColor.cpp` it turns out that the `webkit_color_free()` function calls `WTF::fastFree()`, so this will allocate using the allocator used by the `new` operator, which usually won't be the same used by `WTF::fastFree` when the `GValue` that takes ownership of the boxed `WebKitColor` value destroys it.

So what we want here is `WTF::fastMalloc()`. It would be nicer to have a `webkit_color_new()` helper function, but I think we rather leave that for a separate patch, instead let's do this:

```suggestion
        auto* color = static_cast<WebKitColor*>(fastMalloc(sizeof(WebKitColor)));
```

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -1694,6 +1705,19 @@ static void webkit_web_view_class_init(WebKitWebViewClass* webViewClass)
+     *
+     * The theme color of the WebView's current page.
+     *
+     * Since: TBD
```
The next major release series will be 2.48, therefore:

```suggestion
     * Since: 2.48
```

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -5745,3 +5769,22 @@ webkit_web_view_get_default_content_security_policy(WebKitWebView* webView)
+ * @web_view: a #WebKitWebView
+ * @color: (out): a #WebKitColor to fill in with the theme color
+ *
+ * Gets the theme color that is specified by the content it the @web_view.
```
Is the “it” a typo? I suppose you wanted to use “in”:

```suggestion
 * Gets the theme color that is specified by the content in the @web_view.
```

aperezdc commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -5009,6 +5009,10 @@ void WebPage::updateRendering()
     m_page->updateRendering();
 
+#if PLATFORM(WPE)
+    flushPendingThemeColorChange();
```
Is this not needed for the GTK port?

aperezdc replied:
I went to double check and this should be moved into `DrawingAreaCoordinatedGraphics::display()`, next to the call to `flushPendingEditorStateUpdate()`. All the other ports do such updates in their DrawingArea implementation. Once moved, then the call does not need to be guarded with `PLATFORM(GTK) || PLATFORM(WPE)`.

fredldotme replied:
I've tried this and once moved it didn't flush the theme color state, the theme color stayed invalid. Only from here would it actually do the right thing.

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -5745,3 +5769,22 @@ webkit_web_view_get_default_content_security_policy(WebKitWebView* webView)
+ *
+ * Gets the theme color that is specified by the content it the @web_view.
+ *
+ * Since: tbd
```
Same as above:

```suggestion
 * Since: 2.48
```

aperezdc reviewed (changes requested):
@fredldotme Thanks for sending this, patch looks quite good already. 

I left an important comment about memory allocation&mdash;kind of expected because we have an “interesting” mixture of memory methods (system allocator via GLib, WebKit's bmalloc/fastMalloc/libpas which sometimes might fallback to the system allocator, and the C++ standard `new`/`delete`), but you managed to have the rest of things spot-on 👍🏼 

Could you please go through the comments and update the patch? I appreciate you taking the time.

aperezdc reviewed (changes requested):
@fredldotme Just one more small change request, I went to check how other ports do things and didn't notice one thing before.

Could you also add the `[GTK]` tag in the commit log summary line, please?

My apologies for making you go through another round of changes 🙏🏼

aperezdc commented on Source/WebKit/UIProcess/API/wpe/WebKitWebViewWPE.cpp:
```diff
@@ -326,3 +326,22 @@ void webkit_web_view_toggle_inspector(WebKitWebView* webView)
+    g_return_if_fail(WEBKIT_IS_WEB_VIEW(webView));
+    auto& page = webkitWebViewGetPage(webView);
+
+    g_return_if_fail(page.themeColor().isValid());
```
Checking the implementation of `Page::themeColor()` and `Document::themeColor()` I see that there are two cases in which the color may end up being invalid:

- There is no top-level `Document` (for example: may happen during loads, temporarily)
- The theme color cannot be determined (not declared with the corresponding `<meta>` element, no app-manifest, etc.)

These are cases that we should handle. The `Color` class has a comment telling that it can represent a *Special "invalid color" state, treated as transparent black but distinguishable*.

Taking the above into account, I think this function should return `gboolean` to indicate whether the color is valid. In case of getting an invalid color, fill the `WebKitColor`, with `red = green = blue = alpha = 0.0`.

aperezdc reviewed (changes requested):
I think this is looking good, but I would like to request one last change, to handle the case in which WebKit cannot determine the page theme color, because that is an acceptable case.

@carlosgcampos @mcatanzaro Would you please take a look to rubber-stamp the patch, and have a second approval, as preferred for API additions?

aperezdc commented on Source/WebKit/UIProcess/API/wpe/WebKitWebViewWPE.cpp:
```diff
@@ -326,3 +326,28 @@ void webkit_web_view_toggle_inspector(WebKitWebView* webView)
+ * @color: (out): a #WebKitColor to fill in with the theme color
+ *
+ * Gets the theme color that is specified by the content in the @web_view.
+ *
```
We are missing here a `Returns:` annotation:

```suggestion
 *
 * Returns: Whether the currently loaded page defines a theme color.
 *
```

fredldotme replied:
Done.

aperezdc reviewed (changes requested):
@fredldotme Just one last little bit: you forgot to document what the returned boolean means.

aperezdc reviewed (approved):
Thanks again for the update @fredldotme, approved!

@carlosgcampos Could you have a look and comment whether the API addition is okay for you as well?

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -1709,6 +1729,21 @@ static void webkit_web_view_class_init(WebKitWebViewClass* webViewClass)
+        "theme-color",
+        nullptr, nullptr,
+        WEBKIT_TYPE_COLOR,
+        static_cast<GParamFlags>(WEBKIT_PARAM_READABLE));
```
```suggestion
        WEBKIT_PARAM_READABLE);
```

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/WebKitWebViewWPE.cpp:
```diff
@@ -326,3 +326,30 @@ void webkit_web_view_toggle_inspector(WebKitWebView* webView)
+ * Since: 2.50
+ */
+gboolean
+webkit_web_view_get_theme_color(WebKitWebView* webView, WebKitColor* color)
```
```suggestion
gboolean webkit_web_view_get_theme_color(WebKitWebView* webView, WebKitColor* color)
```

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/WebKitWebViewWPE.cpp:
```diff
@@ -326,3 +326,30 @@ void webkit_web_view_toggle_inspector(WebKitWebView* webView)
+gboolean
+webkit_web_view_get_theme_color(WebKitWebView* webView, WebKitColor* color)
+{
+    g_return_val_if_fail(WEBKIT_IS_WEB_VIEW(webView), false);
```
color isn't nullable, so we should add a g_return_val_if_fail for the color too. However, I wonder if it should be nullable instead, if you just want to check whether the page defines a theme color.

fredldotme replied:
I suppose that was @aperezdc intention.

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/WebKitWebViewWPE.cpp:
```diff
@@ -326,3 +326,30 @@ void webkit_web_view_toggle_inspector(WebKitWebView* webView)
+    auto& page = webkitWebViewGetPage(webView);
+
+    if (!page.themeColor().isValid()) {
+        WebCore::Color tmpColor(WebCore::Color::transparentBlack);
+        webkitColorFillFromWebCoreColor(tmpColor, color);
```
This should be documented. Why do we fill the color with transparent black? I would expect the color not to be filled at all, but if we want to do this we should say it in the documentation.

fredldotme replied:
Did this per @aperezdc request. Added a note to the docs.

aperezdc replied:
> This should be documented. Why do we fill the color with transparent black? I would expect the color not to be filled at all, but if we want to do this we should say it in the documentation.

Because when you get the property using i.e. `g_object_get()` there is no way of telling the caller “there was't a color defined”, so one needs to fill the returned color with something. WebCore's `Color` class uses transparent black in that case, which IMO it makes sense in this case: if one “sees“ a transparent color, that would not really make much sense for a theme color, so that's kind of a neutral value for an invalid color.

carlosgcampos replied:
Makes sense, let's just document it then.

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/WebKitWebViewWPE.cpp:
```diff
@@ -326,3 +326,30 @@ void webkit_web_view_toggle_inspector(WebKitWebView* webView)
+    if (!page.themeColor().isValid()) {
+        WebCore::Color tmpColor(WebCore::Color::transparentBlack);
+        webkitColorFillFromWebCoreColor(tmpColor, color);
+        return false;
```
```suggestion
        return FALSE;
```

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/WebKitWebViewWPE.cpp:
```diff
@@ -326,3 +326,30 @@ void webkit_web_view_toggle_inspector(WebKitWebView* webView)
+    }
+
+    webkitColorFillFromWebCoreColor(page.themeColor(), color);
+    return true;
```
```suggestion
    return TRUE;
```

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -5045,6 +5045,10 @@ void WebPage::updateRendering()
+#if PLATFORM(WPE)
+    flushPendingThemeColorChange();
+#endif
+
```
Why do we need this here? It's already called by the drawing area right after updateRedering().

fredldotme replied:
I left the second place in for review, this one is the one that works, the one triggered from the drawing area doesn't. Should I remove it from the drawing area?

fredldotme replied:
Left the intact one in there, removed the one from the drawing area. I suppose if this gets ported to GTK you might want to find a better spot to put the flush operation.

carlosgcampos replied:
Then we need to understand why it doesn't work there, maybe we are missing other updateRendering callers.

mcatanzaro replied:
Also, unless you understand why this is not needed for macOS, the guard should be:

```
#if PLATFORM(MAC) || PLATFORM(WPE)
```

mcatanzaro replied:
Well, I see the reason: macOS does this in `TiledCoreAnimationDrawingArea::updateRendering`. Ideally we would parallel that to reduce odds of platform-specific bugs, but I see you've already tried....

carlosgcampos replied:
The parallel in our case would be LayerTreeHost::flushLayers, where it was in a previous version of the patch.

fredldotme replied:
Clarification: Previously it was in `DrawingAreaCoordinatedGraphics.cpp`, but I've since moved the flush operation over to `LayerTreeHost::flushLayers`. Doesn't seem to be problematic.

carlosgcampos reviewed (changes requested):
Thanks for the patch. I have a few minor comments and a question, but the most important thing is that this is new API, so a new API test is required.

carlosgcampos commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebView.cpp:
```diff
@@ -1611,6 +1611,30 @@ static void testWebViewBackgroundColor(WebViewTest* test, gconstpointer)
+static void testWebViewThemeColor(WebViewTest* test, gconstpointer)
+{
+    WebKitColor rgba;
+    webkit_web_view_get_theme_color(test->m_webView.get(), &rgba);
```
```suggestion
    g_assert_false(webkit_web_view_get_theme_color(test->m_webView.get(), &rgba));
```

carlosgcampos commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebView.cpp:
```diff
@@ -1611,6 +1611,30 @@ static void testWebViewBackgroundColor(WebViewTest* test, gconstpointer)
+
+    test->loadHtml("<html style='width: 325px; height: 615px'><meta name='theme-color' content='#FF0000' /></html>", nullptr);
+    test->waitUntilLoadFinished();
+    test->wait(1);
```
I don't think we should wait 1 second if it can be avoided, could we just monitor the property by connecting to the notify signal and start the run loop until the signal is emitted?

carlosgcampos commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebView.cpp:
```diff
@@ -1611,6 +1611,30 @@ static void testWebViewBackgroundColor(WebViewTest* test, gconstpointer)
+    test->loadHtml("<html style='width: 325px; height: 615px'><meta name='theme-color' content='#FF0000' /></html>", nullptr);
+    test->waitUntilLoadFinished();
+    test->wait(1);
+    test->showInWindow();
```
Why do we need this at this point?

carlosgcampos commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebView.cpp:
```diff
@@ -1611,6 +1611,30 @@ static void testWebViewBackgroundColor(WebViewTest* test, gconstpointer)
+{
+    WebKitColor rgba;
+    webkit_web_view_get_theme_color(test->m_webView.get(), &rgba);
+
```
```suggestion
```

carlosgcampos commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebView.cpp:
```diff
@@ -1611,6 +1611,30 @@ static void testWebViewBackgroundColor(WebViewTest* test, gconstpointer)
+    test->wait(1);
+    test->showInWindow();
+
+    webkit_web_view_get_theme_color(test->m_webView.get(), &rgba);
```
```suggestion
    g_assert_true(webkit_web_view_get_theme_color(test->m_webView.get(), &rgba));
```

carlosgcampos commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebView.cpp:
```diff
@@ -1611,6 +1611,30 @@ static void testWebViewBackgroundColor(WebViewTest* test, gconstpointer)
+    g_assert_cmpfloat(rgba.green, ==, 0);
+    g_assert_cmpfloat(rgba.blue, ==, 0);
+    g_assert_cmpfloat(rgba.alpha, ==, 1);
+}
```
We could also check that the theme color changes again if we load a different page with a different theme color and another with no theme color

carlosgcampos commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebView.cpp:
```diff
@@ -1611,6 +1611,60 @@ static void testWebViewBackgroundColor(WebViewTest* test, gconstpointer)
+        g_main_loop_run(m_mainLoop);
+    }
+
+    WebKitColor targetColor {0.0, 0.0, 0.0, 0.0};
```
```suggestion
    WebKitColor targetColor { 0.0, 0.0, 0.0, 0.0 };
```

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -1710,6 +1730,21 @@ static void webkit_web_view_class_init(WebKitWebViewClass* webViewClass)
 
+#if PLATFORM(WPE)
+    /**
+     * WebKitWebView::theme-color:
```
```suggestion
     * WebKitWebView:theme-color:
```

carlosgcampos replied:
I'm sorry I didn't notice this before `::` is used for signals

fredldotme committed:
[WPE] Implement theme color retrieval API

https://bugs.webkit.org/show_bug.cgi?id=273747

Reviewed by Adrian Perez de Castro and Carlos Garcia Campos.

Implements a property "theme-color" and a getter function webkit_web_view_get_theme_color()
for retrieving the web view's page theme color, with property changes for notification.

Combined changes:

* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
(WebKitWebViewClient::themeColorDidChange):
(webkitWebViewGetProperty):
(webkit_web_view_class_init):
* Source/WebKit/UIProcess/API/glib/WebKitWebView.h.in:
* Source/WebKit/UIProcess/API/wpe/APIViewClient.h:
(API::ViewClient::themeColorDidChange):
* Source/WebKit/UIProcess/API/wpe/PageClientImpl.cpp:
(WebKit::PageClientImpl::themeColorDidChange):
* Source/WebKit/UIProcess/API/wpe/PageClientImpl.h:
* Source/WebKit/UIProcess/API/wpe/WPEWebView.cpp:
(WKWPE::View::themeColorDidChange):
* Source/WebKit/UIProcess/API/wpe/WPEWebView.h:
* Source/WebKit/UIProcess/API/wpe/WebKitWebViewClient.h:
* Source/WebKit/UIProcess/API/wpe/WebKitWebViewWPE.cpp:
(webkit_web_view_get_theme_color):
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/LayerTreeHost.cpp:
(WebKit::LayerTreeHost::flushLayers):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebView.cpp:
(ThemeColorWebViewTest::themeColorChanged):
(ThemeColorWebViewTest::testThemeColorResult):
(ThemeColorWebViewTest::waitUntilThemeColorChanged):
(testWebViewThemeColor):
(beforeAll):

Canonical link: https://commits.webkit.org/296035@main
